scripts/add_electricity.py

```diff
@@ -476,7 +481,72 @@ def attach_extendable_generators(n, costs, ppl):
 
+    if tech_map is None:
+        tech_map = (snakemake.config['electricity']
+                    .get('include_renewable_capacities_from_OPSD', {}))
+
+    capacities = (ppm.data.OPSD_VRE_country(country).powerplant.convert_country_to_alpha2()
+                  .astype({'Capacity': 'float'})
+                  .groupby(['lat','lon','Fueltype','Technology']) # reduce input data
+                  .agg({'Capacity': 'sum'})
+                  .query('Capacity > 0.1') # only capacities above 100kW
+                  .reset_index())
+
+    # distinguish between onshore and offshore, if given:
+    capacities.Fueltype = (capacities[['Fueltype','Technology']]
+                           .apply(lambda b: b.Technology if b.Technology in ['Onshore', 'Offshore']
+                                  else b.Fueltype, axis=1))
+
+    # retrieve a mapping ppl->bus if locations (lat/lon) for ppls are given:
+    if ~capacities[['lat','lon']].isna().stack().all():
+        buses = n.buses[n.buses.country==country].index
+        
+        mapping = pd.Series(dtype='float')
+        for ppm_fueltype, techs in tech_map.items():
+            for tech in techs:
+                tech_i = n.generators.query('carrier in @tech and bus in @buses').index
+
+                # fix available buses in map_ppls_to_bus function:
+                map_ppls = partial(map_ppls_to_bus, buses=n.generators.loc[tech_i].bus, tech=tech)
+
+                # parallel assignment of ppls to buses; each process is independent to speed up
+                pool_buses = multiprocessing.Pool(processes=4)
+                mapping = (mapping.append(
+                    pd.Series((capacities.query('Fueltype in @ppm_fueltype').Capacity/len(techs)).values,
+                              (pool_buses.map(map_ppls,
+                                              np.array(capacities.query('Fueltype in @ppm_fueltype')[['lat','lon']]))))
+                    .groupby(level=0).sum())) # aggregate capacities per bus with groupby
+                
+        logger.info('Adding {} renewable assets in {} from OPSD'
+                .format(len(capacities), country))
+        return mapping
+    else: # if locations are not given, fall back to heuristics
+        logger.warning('lat and lon are not given for assets in {}; falling back to heuristics'
+                .format(country))
+        return pd.Series(dtype='float')
+
+def attach_OPSD_renewables(n, countries=None):
+    if countries is None:
+        countries = (snakemake.config['electricity']
+                     .get('OPSD_VRES_countries', {}))
+
+    if len(countries) == 0: return
+
+    for country in countries:
+        mapping = attach_OPSD_ppls_for_country(country)
+        n.generators.loc[mapping.index, 'p_nom'] = mapping
+        logger.info('overwriting renewable capacities in {} [MW] with precise locations\n{}'
+                    .format(country, n.generators.loc[mapping.index].groupby('carrier').sum().p_nom))
```

Comment:
```suggestion
def attach_OPSD_renewables(n):
    available = ['DE', 'FR', 'PL', 'CH', 'DK', 'CZ', 'SE', 'GB']
    countries = set(available) & set(n.buses.country)
    tech_map = (snakemake.config['electricity']
                 .get('include_renewable_capacities_from_OPSD', {}))
    logger.info(f'Using OPSD renewable capacities in {", ".join(countries)}.')

    df = pd.concat([pm.data.OPSD_VRE_country(c) for c in countries])
    df = df.query('Fueltype in @techs').powerplant.convert_country_to_alpha2()
    technology_b = ~df.Technology.isin(['Onshore', 'Offshore'])
    df['Fueltype'] = df.Fueltype.where(technology_b, df.Technology)

    for fueltype, carrier_like in tech_map.items():
        gens = n.generators[lambda df: df.carrier.str.contains(carrier_like)]
        buses = n.buses.loc[gens.bus.unique()]
        gens_per_bus = gens.groupby('bus').p_nom.count()

        caps = map_country_bus(df.query('Fueltype == @fueltype'), buses)
        caps = caps.groupby(['bus']).Capacity.sum()

        n.generators.p_nom.update(gens.bus.map(caps).dropna())
```

Could you try out this function? (uses the current ppm master). It should include all the functionality from above